migrations/2025-07-22-091706_multiple_emails/up.sql

```diff
@@ -0,0 +1,127 @@
+-- Drop the unique constraint on user_id to allow multiple emails per user
+ALTER TABLE emails DROP CONSTRAINT emails_user_id_key;
+
+-- Limit users to 32 emails maximum
+CREATE FUNCTION enforce_max_emails_per_user()
+RETURNS TRIGGER AS $$
+BEGIN
+  IF (SELECT COUNT(*) FROM emails WHERE user_id = NEW.user_id) > 32 THEN
+    RAISE EXCEPTION 'User cannot have more than 32 emails';
+  END IF;
+  RETURN NEW;
+END;
+$$ LANGUAGE plpgsql;
+
+CREATE TRIGGER trigger_enforce_max_emails_per_user
+BEFORE INSERT ON emails
+FOR EACH ROW
+EXECUTE FUNCTION enforce_max_emails_per_user();
+
+-- Add a unique constraint for the combination of user_id and email
+ALTER TABLE emails ADD CONSTRAINT unique_user_email UNIQUE (user_id, email);
+
+-- Add a new column for identifying if an email should receive notifications
+ALTER TABLE emails ADD COLUMN send_notifications BOOLEAN DEFAULT FALSE NOT NULL;
+
+-- Set `send_notifications` to true for existing emails
+UPDATE emails SET send_notifications = true;
+
+-- Limit notification flag to one email per user
+-- Evaluation of the constraint is deferred to the end of the transaction to allow for replacement of the notification email
+CREATE EXTENSION IF NOT EXISTS btree_gist;
+ALTER TABLE emails ADD CONSTRAINT unique_notification_email_per_user
+EXCLUDE USING gist (
+  user_id WITH =,
+  (send_notifications::int) WITH =
+)
+WHERE (send_notifications)
+DEFERRABLE INITIALLY DEFERRED;
+
+-- Prevent deletion of emails if they have notifications enabled, unless it's the only email for that user
+CREATE FUNCTION prevent_notification_email_deletion()
+RETURNS TRIGGER AS $$
+BEGIN
+  IF OLD.send_notifications IS TRUE THEN
+    -- Allow deletion if this is the only email for the user
+    IF (SELECT COUNT(*) FROM emails WHERE user_id = OLD.user_id) = 1 THEN
+      RETURN OLD;
+    END IF;
+    RAISE EXCEPTION 'Cannot delete email: send_notifications is set to true';
+  END IF;
+  RETURN OLD;
+END;
+$$ LANGUAGE plpgsql;
```

Comment:
why do we allow deletion if it's the only email address?

Reply:
You can't delete a user row if there is an email row referencing it, so I did this to allow for users to be deleted.

Whether that's beneficial in development or production is a question to be answered I suppose.

Reply:
Yeah there is a constraint preventing users from being deleted that was introduced in https://github.com/rust-lang/crates.io/pull/1090

```
update or delete on table "users" violates foreign key constraint "fk_emails_user_id" on table "emails"
Detail: Key (id)=(2) is still referenced from table "emails".
```

We can discuss that when the schema changes are split into a separate PR.